governance/node.go

```diff
@@ -80,49 +79,9 @@ func (w *GovernanceWorker) handleNodeUserInputUpdated(svcSpecs persistence.Servi
 
 	glog.V(5).Infof(logString(fmt.Sprintf("handling node user input changes")))
 
-	if svcSpecs == nil || len(svcSpecs) == 0 {
-		return
-	}
-
-	// get all the unarchived agreements
-	agreements, err := persistence.FindEstablishedAgreementsAllProtocols(w.db, policy.AllAgreementProtocols(), []persistence.EAFilter{persistence.UnarchivedEAFilter()})
-	if err != nil {
-		glog.Errorf(logString(fmt.Sprintf("Unable to retrieve all the  from the database, error %v", err)))
-		return
-	}
-
-	// cancel the agreement if needed
-	for _, ag := range agreements {
-		agreementId := ag.CurrentAgreementId
-		if ag.AgreementTerminatedTime != 0 && ag.AgreementForceTerminatedTime == 0 {
-			glog.V(3).Infof(logString(fmt.Sprintf("skip agreement %v, it is already terminating", agreementId)))
-		} else {
-			bCancel, err := w.agreementRequiresService(ag, svcSpecs)
-			if err != nil {
-				glog.Errorf(fmt.Sprintf("%v", err))
-			}
-
-			if bCancel {
-				glog.V(3).Infof(logString(fmt.Sprintf("ending the agreement: %v", agreementId)))
-
-				reason := w.producerPH[ag.AgreementProtocol].GetTerminationCode(producer.TERM_REASON_NODE_USERINPUT_CHANGED)
-
-				eventlog.LogAgreementEvent(
-					w.db,
-					persistence.SEVERITY_INFO,
-					persistence.NewMessageMeta(EL_GOV_START_TERM_AG_WITH_REASON, ag.RunningWorkload.URL, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason)),
-					persistence.EC_CANCEL_AGREEMENT,
-					ag)
-
-				w.cancelAgreement(agreementId, ag.AgreementProtocol, reason, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason))
-
-				// send the event to the container in case it has started the workloads.
-				w.Messages() <- events.NewGovernanceWorkloadCancelationMessage(events.AGREEMENT_ENDED, events.AG_TERMINATED, ag.AgreementProtocol, agreementId, ag.GetDeploymentConfig())
-				// clean up microservice instances if needed
-				w.handleMicroserviceInstForAgEnded(agreementId, false)
-			}
-		}
-	}
+	// Make sure that every agreement we have is in a valid state or proceeding to valid states in a timely fashion. If not,
+	// cancel the agreement and allow the agbots to re-make them if necessary.
+	w.governAgreements()
```

Comment:
governAgreements check the compatibility of of the policies. It does not handle the user input changes.

Reply:
Tracing the logs (generated by an IEAM 4.3 agent v2.29.0), the logic flows and error messages point to this function and causes the agreements to be cancelled.

(I see a PR from 13 months ago that changed the behavior of all of this.  Is the 2.29 branch that much older than 2.30 ?)  It's possible I was tracing the logs generated from a 2.29.0 agent and the 2.30.0 code base is substantially divergent. 

https://github.com/open-horizon/anax/commit/2b0219fbfac1ef378c86b75c4750860a68e539cf#diff-8f7cb8d1fb0d5b06387d5655c337658bd5ff41ef9317f1499442b7683b55a693L47

Reply:
One way to test it is:
1. start e2edev test.
2. hzn userinput list > /tmp/ui.json
3.  change a value in /tmp/ui.json
4. hzn userinput add -f /tmp/ui.json  see if the agreement got canceled or not. The agreements that uses the value are supposed to be canceled.